doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,551 @@
+  delaying completion.
+
+
+Crash consistency
+------------------
+* Large case
+  1. Crash occurs right after writing Data blocks
+
+    - Data partition --> | Data blocks |
+    - We don't need to care this case. Data is not alloacted yet in reality. The blocks will be reused.
+
+  2. Crash occurs right after WAL 
+
+    - Data partition --> | Data blocks |
+    - WAL --> | TxBegin A | Log Entry | TxEnd A |
+    - Write procedure is completed, so there is no data loss or inconsistent state
+
+* Small case
+  1. Crash occurs right after writing WAL
+
+    - WAL --> | TxBegin A | Log Entry| TxEnd A |
+    - All data has been written
+
+
+Comparison
+----------
+* Best case (pre-allocation)
+  - Only need writes on both WAL and Data partition without updating object metadata (for the location).
+* Worst case 
+  - At least three writes are required additionally on WAL, object metadata, and data blocks.
+  - If the flush from WAL to the data parition occurs frequently, radix tree onode structure needs to be update
+  in many times. To minimize such overhead, we can make use of batch processing to minimize the update on the tree
+  (the data related to the object has a locality because it will have the same parent node, so updates can be minimized)
+
+* WAL needs to be flushed if the WAL is close to full or a signal to flush.
+  - The premise behind this design is OSD can manage the latest metadata as a single copy. So,
+  appended entries are not to be read
+* Either best of the worst case does not produce severe I/O amplification (it produce I/Os, but I/O rate is constant) 
+  unlike LSM-tree DB (the proposed design is similar to LSM-tree which has only level-0)
+  
+
+Detailed Design 
+===============
+
+* Onode lookup
+
+  * Radix tree
+    Our design is entirely based on the prefix tree. Ceph already makes use of the characteristic of OID's prefix to split or search
+    the OID (e.g., pool id + hash + oid). So, the prefix tree fits well to store or search the object. Our scheme is designed 
+    to lookup the prefix tree efficiently.
+
+  * Sharded partition
+    A few bits (leftmost bits of the hash) of the OID determine a sharded partition where the object is located.
+    For example, if the number of partitions is configured as four, The entire space of the hash in hobject_t 
+    can be divided into four domains (0x0xxx ~ 0x3xxx, 0x4xxx ~ 0x7xxx, 0x8xxx ~ 0xBxxx and 0xCxxx ~ 0xFxxx).  
+
+  * Ondisk onode ::
+
+            stuct onode {
+              onode childs;
+              onode parent_node;
+              extent_tree block_maps;
+              b+_tree omaps;
+            }
+
+    onode contains the radix tree nodes for lookup, which means we can search for objects using tree node information in onode. 
+    Also, if the data size is small, the onode can embed the data.
+    The onode is fixed size (256 or 512 byte). On the other hands, omaps and block_maps are variable-length by using pointers in the onode. ::
```

Comment:
I continue to be confused by this part.  I assume that the point of an onode radix tree is to index onodes by ghobject_t.  From the fact that you have 'onode childs' and 'onode parent_node' I guess that the internal nodes of your tree for some reason contain onodes.  Why?  Especially with a radix tree, one generally chooses to have only keys in internal nodes to reduce fan-out with the actual values (onodes, here) only in leaf nodes.

Also, where do you plan on storing the xattrs?

Reply:
Sorry for confusing. Onode here is an onode pointer to indicate another onode. Like xa_node in linux kernel, parent is an 
onode pointer to root, and childs is a pointer array the onode belongs to.

I think the omap structure is able to handle the xattrs because they have the same data model where each key is associated with the value. Do I need a separated data structure for xattrs?

Reply:
I think you are describing a basic in-memory n-ary tree implementation.  A radix tree as you insist that you are using here is quite different, particularly when implemented on disk.  For the purposes of this discussion, you probably want to remove references to the indexing structures from the onode itself -- they are orthogonal (or at least mostly so).

You stated that the onode is fixed size.  I suggest that you store the xattrs within the onode and permit it to be variably sized -- if only because the ghobject key already is.

Reply:
I got it.